#164691235 Article readtime intergration.

NicholusMuwonge opened the pull request:
#### What does the PR do?
Alerts the reader of how much time they will need to read a certain article.

#### Description of the task to be completed?
- The feature should count 256 words and assign a minute to be read according to Medium's average reading time of most people.
- All the read time is displayed in minutes and rounded off to whole numbers
- All read times below one minute are automatically assigned "1 min read"

#### How should this be manually tested?
- Clone this repository locally and checkout to ```ft-article-readtime-164691235``` 
- Carry out the necessary setting up the database, environments , install all requirements .
- ```Pip install readtime``` and run ```py manage.py migrate```
- Start  server and open up postman or any other tool of your choice to test the functionlaity of this feature.

Use test route to create article ```http://localhost:8000/api/articles/```
## test body can look like 
```
{
    "article": {
        "title": "Lorem ipsum dolor ",
        "description": "Lorem ipsum dolor sit amet, consectetur ",
        "body": "Ask Question I added validate_mail and also installed py3dns to check if the email the user signs up with is valid(like their emails are not fake). The tests seem Ask Question I added validate_mail and also installed py3dns to check if the email the user signs up with is valid(like their emails are not fake). The tests seem Ask Question I added validate_mail and also installed py3dns to check if the email the user signs up with is valid(like their emails are not fake). The tests seem Ask Question I added validate_mail and also installed py3dns to check if the email the user signs up with is valid(like their emails are not fake). The tests seem Ask Question I added validate_mail and also installed py3dns to check if the email the user signs up with is valid(like their emails are not fake). The tests seem v Ask Question I added validate_mail and also installed py3dns to check if the email the user signs up with is valid(like their emails are not fake). The tests seem     Ask Question I added validate_mail and also installed py3dns to check if the email the user signs up with is valid(like their emails are not fake). The tests seem vvvv Ask Question I added validate_mail and also installed py3dns to check if the email the user signs up with is valid(like their emails are not fake). The tests seem            Ask Question I added validate_mail and also installed py3dns to check if the email the user signs up with is valid(like their emails are not fake). The tests seem                                                                                                                                                                                                                                        Ask Question I added validate_mail and also installed py3dns to check if the email the user signs up with is valid(like their emails are not fake). The tests seem Ask Question I added validate_mail and also installed py3dns to check if the email the user signs up with is valid(like their emails are not fake). The tests seem  Ask Question I added validate_mail and also installed py3dns to check if the email the user signs up with is valid(like their emails are not fake). The tests seem Ask Question I added validate_mail and also installed py3dns to check if the email the user signs up with is valid(like their emails are not fake). The tests seem Ask Question I added validate_mail and also installed py3dns to check if the email the user signs up with is valid(like their emails are not fake). The tests seem v Ask Question I added validate_mail and also installed py3dns to check if the email the user signs up with is valid(like their emails are not fake). The tests seem Ask Question I added validate_mail and also installed py3dns to check if the email the user signs up with is valid(like their emails are not fake). The tests seemv Ask Question I added validate_mail and also installed py3dns to check if the email the user signs up with is valid(like their emails are not fake). The tests seem "
    }
}
```

and response will look like 
```
{
    "article": {
        "title": "Lorem ipsum dolor",
        "description": "Lorem ipsum dolor sit amet, consectetur",
        "body": "Ask Question I added validate_mail and also installed py3dns to check if the email the user signs up with is valid(like their emails are not fake). The tests seem Ask Question I added validate_mail and also installed py3dns to check if the email the user signs up with is valid(like their emails are not fake). The tests seem Ask Question I added validate_mail and also installed py3dns to check if the email the user signs up with is valid(like their emails are not fake). The tests seem Ask Question I added validate_mail and also installed py3dns to check if the email the user signs up with is valid(like their emails are not fake). The tests seem Ask Question I added validate_mail and also installed py3dns to check if the email the user signs up with is valid(like their emails are not fake). The tests seem v Ask Question I added validate_mail and also installed py3dns to check if the email the user signs up with is valid(like their emails are not fake). The tests seem     Ask Question I added validate_mail and also installed py3dns to check if the email the user signs up with is valid(like their emails are not fake). The tests seem vvvv Ask Question I added validate_mail and also installed py3dns to check if the email the user signs up with is valid(like their emails are not fake). The tests seem            Ask Question I added validate_mail and also installed py3dns to check if the email the user signs up with is valid(like their emails are not fake). The tests seem                                                                                                                                                                                                                                        Ask Question I added validate_mail and also installed py3dns to check if the email the user signs up with is valid(like their emails are not fake). The tests seem Ask Question I added validate_mail and also installed py3dns to check if the email the user signs up with is valid(like their emails are not fake). The tests seem  Ask Question I added validate_mail and also installed py3dns to check if the email the user signs up with is valid(like their emails are not fake). The tests seem Ask Question I added validate_mail and also installed py3dns to check if the email the user signs up with is valid(like their emails are not fake). The tests seem Ask Question I added validate_mail and also installed py3dns to check if the email the user signs up with is valid(like their emails are not fake). The tests seem v Ask Question I added validate_mail and also installed py3dns to check if the email the user signs up with is valid(like their emails are not fake). The tests seem Ask Question I added validate_mail and also installed py3dns to check if the email the user signs up with is valid(like their emails are not fake). The tests seemv Ask Question I added validate_mail and also installed py3dns to check if the email the user signs up with is valid(like their emails are not fake). The tests seem",
        "slug": "lorem-ipsum-dolor-36",
        "createdAt": "2019-04-09 20:54:13.418823+00:00",
        "updatedAt": "2019-04-09 20:54:13.418823+00:00",
        "favorited": false,
        "favoritesCount": 0,
        "read_time": "2 min read",
        "author": {
            "username": "nicksbros"
        }
    }
}

```

#### What are the relevant pivotal tracker stories?
https://www.pivotaltracker.com/story/show/164691235

richien commented on authors/apps/articles/serializers.py:
```diff
@@ -18,14 +19,17 @@ class CreateArticleSerializer(serializers.ModelSerializer):
 
     def create(self, validated_data):
+        read_time = validated_data.pop('read_time',None)
+        read_time = readtime.of_text(validated_data.get('body'),wpm = 265)
```
Hi @NicholusMuwonge , this is good work. Well done. I just have a few things I've noticed with your implementation. Firstly, line 31 of this file is redundant because at this point in the creation of an article, the read_time is not yet in the **validated_data** dictionary. That dictionary contains the data returned from the **validate** method. Therefore this `read_time = validated_data.pop('read_time',None)` will always evaluate to `read_time = None`. In any case, I see that you are initializing the `read_time` on line 32, `read_time = readtime.of_text(validated_data.get('body'),wpm = 265)`, so I am wondering what the purpose of line 31 is.

NicholusMuwonge replied:
hey @richien I am still struggling a bit with django semantics. I see it works fine without line 31 and I have fixed it. Thanks for the pointer

richien reviewed (approved):
I have tested this locally and it functions as you have described. Good work.

RonKbS commented on authors/apps/articles/serializers.py:
```diff
@@ -18,14 +19,16 @@ class CreateArticleSerializer(serializers.ModelSerializer):
         return data
 
     def create(self, validated_data):
+        read_time = readtime.of_text(validated_data.get('body'),wpm = 265)
```
@NicholusMuwonge, great work.

Kindly share where you got the wpm of 265.
Also look into whether this is representative by researching on it.
Please do this ASAP so I can merge this so we can have @richien's PR merged as well.

NicholusMuwonge replied:
I got 256 wpm from medium support center https://help.medium.com/hc/en-us/articles/214991667-Read-time

RonKbS commented on authors/apps/articles/tests/test_create_article.py:
```diff
@@ -41,3 +43,25 @@ def test_api_cannot_create_an_article_with_an_unauthenticated_user(self):
+        )
+
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
+        self.assertIn('2 min read', response.data["read_time"])
```
@richien, @NicholusMuwonge, @Manorlds-Eaglespark and @reifred,
I believe we need a check such that builds will fail if we do not follow standards.
Kindly look into how to use flake8 0r pep8 so PR tests do not pass if we have issues such as the missing line here.

richien replied:
Ok @RonKbS, I am working on adding flake8 to the branch I am currently working on.

RonKbS replied:
I applaud your proactiveness in doing so @richien. But we will have it as a separate story for next week. For now let us go ahead without them. Same goes for you, @NicholusMuwonge. Please remove the flake8 implementation from this PR, as well as the changes I see related to creating profiles
<img width="1066" alt="Screenshot 2019-04-11 at 14 15 37" src="https://user-images.githubusercontent.com/40039818/55953255-4d026b80-5c64-11e9-86c0-31d29214b333.png">

NicholusMuwonge replied:
I did remove the implementation

RonKbS commented on authors/apps/articles/tests/test_data.py:
```diff
@@ -18,3 +18,9 @@
+valid_article_2= {
+    "title": "How to train your dragon",
+    "description": "Ever wonder how?",
+    "body": "Ask Question I added validate_mail and also installed py3dns to check if the email the user signs up with is valid(like their emails are not fake). The tests seem Ask Question I added validate_mail and also installed py3dns to check if the email the user signs up with is valid(like their emails are not fake). The tests seem Ask Question I added validate_mail and also installed py3dns ...
```
Please break this body into several lines.

NicholusMuwonge replied:
I have fixed this please

RonKbS reviewed (changes requested):
Kindly work on these comments ASAP while you're rebasing.

NicholusMuwonge committed:
feat(readtime): user gets to know article readtime

- the article on being created,automatically displays readtime
- readtime is based on 256 wpm

[Finishes #164691235]

RonKbS commented:
The report generated for this PR is including changes not related to this branch.
Please refrain from changing settings on coveralls, travis, heroku and github without notifying me first, @richien, @NicholusMuwonge, @Manorlds-Eaglespark and @reifred.

If this is because of something else, kindly see if you can find out why.

RonKbS commented on authors/apps/articles/tests/test_create_article.py:
```diff
@@ -41,3 +43,25 @@ def test_api_cannot_create_an_article_with_an_unauthenticated_user(self):
 
         self.assertEqual(response.status_code, status.HTTP_403_FORBIDDEN)
+
+    def test_that_every_256_words_or_below_is_1_min_read(self):
```
I see here that you have 256 yet in the implementation it is 265. Kindly elaborate on which it is and fix accordingly

NicholusMuwonge replied:
Its 256 I had implemented it with a mistake

RonKbS commented on .travis.yml:
```diff
@@ -2,7 +2,9 @@ language: python
 python: 
 - "3.6.6"
-
+matrix:
```
Please share the reason for adding this to the .travis file.

NicholusMuwonge replied:
Oh I was trying to add Flake 8, let me remove it, please

RonKbS reviewed (changes requested):
Ensure that you remove the flake8 implementation in addition to these changes.
